include/tvm/operation.h

```diff
@@ -191,15 +191,36 @@ class TVM_DLL ComputeOpNode : public OperationNode {
   Array<IterVar> axis;
   /*! \brief IterVar on each reduction axis, if the body is a Reduce */
   Array<IterVar> reduce_axis;
+  // override functions
+  Array<IterVar> root_iter_vars() const final;
+  Array<Expr> output_shape(size_t idx) const final;
+  void GatherBound(
+          const Operation& self,
+          const std::unordered_map<Tensor, TensorDom>& tensor_dom,
+          std::unordered_map<IterVar, Range>* out_dom_map) const final;
```

Comment:
Since ComputeOpNode is an exposed structure, can we keep it as the original, but we have another BasicComputeOpNode; ComputeOpNode and TensorComputeOpNode inherit it. in this way, we can avoid so many modifications, also we can save modifications for those in-house projects based on tvm.

Reply:
The way I see it is that TVM has defined a `Node` class, then derived `OperationNode`, then derived `<X>OpNode`, and so on. Therefore, before this PR, one can find out the class hierarchy only by looking at the names of the classes. I believe this was a good decision and I have seen it in other code bases. That's why I thought to keep it that way after this new class is introduced.

Regarding the amount of modifications needed both inside and outside TVM:
* inside TVM: I agree that quite a few files have been affected. However, the changes are trivial, and have already been done fairly quickly using automatic refactoring tools so no additional effort is needed.
* outside TVM: even if there are a lot of code out there using `ComputeOpNode` (which need to be renamed to `ScalarComputeOpNode`), I don't see why it's an inconvenient thing to do that. Maybe I missing something? 
Moreover, I think keeping the consistency of class naming in TVM (which I hope strives for many years to come) is more important than that level of inconvenience in external projects that depend on it. 

That's my two cents.

Reply:
> The way I see it is that TVM has defined a `Node` class, then derived `OperationNode`, then derived `<X>OpNode`, and so on. Therefore, before this PR, one can find out the class hierarchy only by looking at the names of the classes. I believe this was a good decision and I have seen it in other code bases. That's why I thought to keep it that way after this new class is introduced.
> 

agree. what i'm suggesting has no conflicts with this point.

> Regarding the amount of modifications needed both inside and outside TVM:
> 
> * inside TVM: I agree that quite a few files have been affected. However, the changes are trivial, and have already been done fairly quickly using automatic refactoring tools so no additional effort is needed.
> * outside TVM: even if there are a lot of code out there using `ComputeOpNode` (which need to be renamed to `ScalarComputeOpNode`), I don't see why it's an inconvenient thing to do that. Maybe I missing something?

We do have some custmized pass(C++ and python) and costmized schedule template(python) , in which we referenced ComputeOpNode, if this is merged, we need to replace them with ScalarComputeOpNode when sync with upstream. I believe other inhouse projects may have the similar issue with this.

>   Moreover, I think keeping the consistency of class naming in TVM (which I hope strives for many years to come) is more important than that level of inconvenience in external projects that depend on it.

I don't see much differences if we change ComputeOpNode to ScalarComputeOpNode, since everybody who uses `ComputeOpNode` knew it's a scalar compute,  for those who doesn't know, a comment on ComputeOpNode class will be enough.

> 
> That's my two cents.

Reply:
Thanks @xqdan. I followed the naming scheme you recommended. I used `BaseComputeOpNode` for the base class.